Enable partitioning for temporary tables with Presto + Hive

dongyoungy opened the pull request:
1. Enable partitioning for temporary tables with Presto + Hive
2. Add support for partial scramble table with SIZE (e.g., CREATE SCRAMBLE A FROM B SIZE 0.2)
3. Add support for specifying block size for scramble table with BLOCKSIZE (e.g., CREATE SCRAMBLE A FROM B BLOCKSIZE 10000)
4. Fix a bug ScrambleMetaStore retrieves the oldest meta information for scrambles instead of most recent one

I will update the documentation once this gets approved..

@pyongjoo I also verified that nodes get executed sequentially in order of their block numbers. Adding a delay between `runOnThreads()` gave a worse performance so I did not touch them.

dongyoungy commented:
@pyongjoo I have pushed changes addressing your comments. Note that the query writing for SIZE happens in CreateTableToSql. Please take a look and let me know if there are other issues.

BTW, our temporary Presto test instance in Google Cloud is currently down because my free trial credit ran out ($300 in 2 weeks!). This is caused by what seems to be an excessive amount of network usage, which I do not know why and how it happened. I will contact their support to figure this out later. 

Nonetheless, I think we need to move/configure our Presto test instance in one of our lab machines as we discussed previously, now knowing that we do not need to have one in Google Cloud.

Since I am traveling tonight, I will try to set it up tomorrow and let you know.

pyongjoo commented:
Thanks for the update. I’ll take a look. 

Regarding the lab machine, salat2 will be good I guess since you also have sudo access.

pyongjoo commented on src/main/java/org/verdictdb/coordinator/ScramblingCoordinator.java:
```diff
@@ -40,7 +33,13 @@
+import java.util.Map.Entry;
+import java.util.Set;
 
 public class ScramblingCoordinator {
```
The current logic first calculates the number of blocks from the original table size; then, reduce the number of blocks if the relative size is set to smaller than 1.0. This approach will be less optimal when the relative size is small (e.g., 0.01).

An alternative way we can do is
```java
int effectiveTableRowCount = tableRowCount * relativeSize;
int blockCount = min(maxScrambleTableBlockCount, ceil(effectiveTableRowCount / blockSize));
double blockSize = effectiveTableRowCount / (double) blockCount;
int blockCountUpperBound = ceil(tableRowCount / blockSize);
```
Then, use sql query as follows:
```sql
CREATE TABLE scramble AS
SELECT * FROM (
  SELECT *, floor(rand() * blockCountUpperBound) as verdictdbblock 
  FROM originalTable
) t
WHERE verdictdbblock < blockCount
```

dongyoungy replied:
I guess your point is that the current approach may create a scramble table larger than specified size. However, in other words, the current approach guarantees at least 1% (or 1 million rows) of original data gets included in the scramble table with 'maxBlockCount=100', but I think your proposed approach has a potential problem of allowing a scramble table having too few rows? 

Think about an extreme case where `effectiveTableRowCount` becomes 1 with `relativeSize` that is really small (like 0.00000001). We will end up with one single-row block for the scramble table (assuming we also guard the case where blockSize becomes 0) that is not going to work for any approximation. I guess it comes down to a decision about how much we let users to control the size of their data with this command.  Wdyt?

pyongjoo replied:
Hmm. Your example confuses me. If the user wants, say, 1e-6 fraction of the original rows to be included in a scramble, why should we stop it?

Let me give you an example to clarify what I meant. Suppose that the original table has 1B rows and the relative size is 0.01 (thus, the effective size is 1B*0.01=10M). Then, I think we should create 10 blocks (each with size 1M) rather than creating 1 block with 10M rows.

However, since the number of block count limit is 100, the current logic first set the block count to 100 where each block contains 1B/100 = 10M rows. Then, since the relative size is 0.01, only the first block (of size 10M) will be created.

dongyoungy replied:
> say, 1e-6 fraction of the original rows to be included in a scramble, why should we stop it?
My question was about allowing/preventing users to create scrambles that may have no approximation power at all. Why do we allow users to create scrambles that will be useless anyway? I am just saying that there could be such case for the record. 

I will make the change anyway as I agree on the other point about having 1M * 10 blocks rather than 10M * 1 block.

pyongjoo commented:
The comment I just made is the only comment.

pyongjoo commented:
Let me check if we have any backup snapshots.

pyongjoo commented:
@dongyoungy I attach the restored cluster address. It has "tpch10g" schema. Let me know after updating CircleCI setting. I will kill the old cluster.

`verdictdb-test2.cquckbw8fkl6.us-east-1.redshift.amazonaws.com`

dongyoungy commented:
@pyongjoo Thank you for fixing the issue. However, it looks like the problem still persists even after I dropped all garbage schemas in the new instance. It would be great if you can take a look at it. The following is the error message from one of the unit tests involving the redshift test instance (CircleCI environmental variables have been updated for the new instance):

```
[Amazon](500310) Invalid operation: Disk Full
Details: 
 -----------------------------------------------
  error:  Disk Full
  code:      1016
  context:   node: 0
  query:     3182
  location:  fdisk_api.cpp:588
  process:   query0_124_3182 [pid=8941]
  -----------------------------------------------;
```

BTW, Presto test instance has been setup @ salat2 (i.e., salat2.eecs.umich.edu:9080) and I have addressed your latest comments in this pull request so please take a look at it afterwards.

pyongjoo commented:
I just dropped verdictdbtemp schema, which may have resolved the issue. This is the result of a query that asks free space.
```
dev=# select
dev-#   sum(capacity)/1024 as capacity_gbytes, 
dev-#   sum(used)/1024 as used_gbytes, 
dev-#   (sum(capacity) - sum(used))/1024 as free_gbytes 
dev-# from 
dev-#   stv_partitions where part_begin=0;
 capacity_gbytes | used_gbytes | free_gbytes 
-----------------+-------------+-------------
             353 |          26 |         327
(1 row)
```

dongyoungy commented:
I forgot to leave a comment, but changes have been made to address all issues raised so far in this thread.

pyongjoo commented on src/main/java/org/verdictdb/coordinator/ScramblingCoordinator.java:
```diff
@@ -44,17 +44,13 @@
 // as follows:
 
-// 1. blockCount = ceil(tableRowCount / blockSize)
+// 1. scrambleTableSize = ceil(tableRowCount / relativeSize)
```
I guess it should be multiplication.

pyongjoo commented:
@dongyoungy Thanks. It looks good (although I left a minor comment). Please add documentation after merging this.
